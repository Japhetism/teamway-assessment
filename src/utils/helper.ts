import { questions } from "../constants/questions"

export const getQuestionHelper = (questionNumber: number) => {
    if (questions[questionNumber - 1]) {
        return {
            details: questions[questionNumber - 1],
            questionNumber,
            totalQuestion: questions.length
        }
    } else {
        return {
            error: "Invalid question number",
            questionNumber,
            totalQuestion: questions.length
        }
    }
    
}

export const getButtonText = (questionNumber: number, totalQuestions: number) => {
    return questionNumber === totalQuestions ? "Finish Test" : "Next Question";
}

export const getQuestionAnswerType = (answer: string, extrovertOptions: string[], introvertOptions: string[]) => {
    if (extrovertOptions.indexOf(answer)) {
        return "Extrovert";
    } else if (introvertOptions.indexOf(answer)) {
        return "Introvert";
    } else {
        return "";
    }
}