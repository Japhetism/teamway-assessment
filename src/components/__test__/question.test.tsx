import { render, within } from "@testing-library/react";
import { createRenderer } from "react-test-renderer/shallow";
import Question from "../question";
import { getQuestionHelper } from "../../utils/helper";

const renderer = createRenderer();

describe("Question Component", () => {

    it("should render and match the snapshot", () => {
        const { details, questionNumber, totalQuestion } = getQuestionHelper(1);
        renderer.render(
            <Question
                question={details?.question}
                options={details?.options}
                questionNumber={questionNumber}
                totalQuestions={totalQuestion}
                onBtnClick={() => {}}
            />
        );
        const renderedOutput = renderer.getRenderOutput();

        expect(renderedOutput).toMatchSnapshot();
    });

    it("should show next button", () => {
        const { details, questionNumber, totalQuestion } = getQuestionHelper(1);
        const { getByRole} = render(
            <Question
                question={details?.question}
                options={details?.options}
                questionNumber={questionNumber}
                totalQuestions={totalQuestion}
                onBtnClick={() => {}}
            />
        );

        const nextBtn = getByRole("button", { name: "Next Question"});

        expect(nextBtn).toBeVisible();
    });

    it("should show finish button if on the last question", () => {
        const { details, questionNumber, totalQuestion } = getQuestionHelper(14);
        const { getByRole} = render(
            <Question
                question={details?.question}
                options={details?.options}
                questionNumber={questionNumber}
                totalQuestions={totalQuestion}
                onBtnClick={() => {}}
            />
        );

        const nextBtn = getByRole("button", { name: "Finish Test"});

        expect(nextBtn).toBeVisible();
    });

    it("should have error component within question component", () => {
        const { error, details, questionNumber, totalQuestion } = getQuestionHelper(17);
        const { getByTestId } = render(
            <Question
                question={details?.question}
                options={details?.options}
                questionNumber={questionNumber}
                totalQuestions={totalQuestion}
                error={error}
                onBtnClick={() => {}}
            />
        );

        const questionComponent = getByTestId("question-component");
        const errCompWithinQuestComp = within(questionComponent).getByTestId("error-component");

        expect(errCompWithinQuestComp).toBeVisible();
    });
});
