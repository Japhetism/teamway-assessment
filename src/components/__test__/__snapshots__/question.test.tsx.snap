// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Component should render and match the snapshot 1`] = `
<div
  className="max-w-3xl rounded overflow-hidden shadow-lg bg-white mt-10"
  data-testid="question-component"
>
  <React.Fragment>
    <div
      className="px-6 py-4"
    >
      <p>
        Question 1/14
      </p>
      <div
        className="font-bold text-xl mb-2 mt-10"
      >
        You’re really busy at work and a colleague is telling you their life story and personal woes. You:
      </div>
      <p>
        <small>
          All questions are required
        </small>
      </p>
      <div
        className="mt-10"
      >
        <div
          className="flex items-center mb-4"
        >
          <input
            className="w-4 h-4 text-black-600 bg-gray-100 dark:bg-gray-700 dark:border-gray-600"
            id="default-radio-a"
            name="default-radio"
            onChange={[Function]}
            type="radio"
            value="a"
          />
          <label
            className="ml-2 text-sm font-medium text-gray-900 dark:text-gray-300"
            htmlFor="default-radio-a"
          >
            Don’t dare to interrupt them
          </label>
        </div>
        <div
          className="flex items-center mb-4"
        >
          <input
            className="w-4 h-4 text-black-600 bg-gray-100 dark:bg-gray-700 dark:border-gray-600"
            id="default-radio-b"
            name="default-radio"
            onChange={[Function]}
            type="radio"
            value="b"
          />
          <label
            className="ml-2 text-sm font-medium text-gray-900 dark:text-gray-300"
            htmlFor="default-radio-b"
          >
            Think it’s more important to give them some of your time; work can wait
          </label>
        </div>
        <div
          className="flex items-center mb-4"
        >
          <input
            className="w-4 h-4 text-black-600 bg-gray-100 dark:bg-gray-700 dark:border-gray-600"
            id="default-radio-c"
            name="default-radio"
            onChange={[Function]}
            type="radio"
            value="c"
          />
          <label
            className="ml-2 text-sm font-medium text-gray-900 dark:text-gray-300"
            htmlFor="default-radio-c"
          >
            Listen, but with only with half an ear
          </label>
        </div>
        <div
          className="flex items-center mb-4"
        >
          <input
            className="w-4 h-4 text-black-600 bg-gray-100 dark:bg-gray-700 dark:border-gray-600"
            id="default-radio-d"
            name="default-radio"
            onChange={[Function]}
            type="radio"
            value="d"
          />
          <label
            className="ml-2 text-sm font-medium text-gray-900 dark:text-gray-300"
            htmlFor="default-radio-d"
          >
            Interrupt and explain that you are really busy at the moment
          </label>
        </div>
      </div>
    </div>
    <div
      className="px-6 pt-4 pb-2"
    >
      <button
        className="text-white bg-gradient-to-r from-blue-500 via-blue-600 to-blue-700 hover:bg-gradient-to-br focus:ring-4 focus:outline-none focus:ring-blue-300 dark:focus:ring-blue-800 shadow-lg shadow-blue-500/50 dark:shadow-lg dark:shadow-blue-800/80 font-medium rounded-lg text-sm px-5 py-2.5 text-center mr-2 mb-2"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Next Question
      </button>
    </div>
  </React.Fragment>
</div>
`;
